File: 05-behavioral-patterns/03-command/03-command-undo.py
from abc import ABC, abstractmethod

# Command Interface
class Command(ABC):
    @abstractmethod
    def execute(self):
        pass

    @abstractmethod
    def undo(self):
        pass

# Receiver
class Light:
    def on(self):
        print("Light is ON")

    def off(self):
        print("Light is OFF")

# Concrete Commands
class LightOnCommand(Command):
    def __init__(self, light: Light):
        self.light = light

    def execute(self):
        self.light.on()

    def undo(self):
        self.light.off()

class LightOffCommand(Command):
    def __init__(self, light: Light):
        self.light = light

    def execute(self):
        self.light.off()

    def undo(self):
        self.light.on()

# Invoker
class RemoteControl:
    def __init__(self):
        self.history = []

    def execute_command(self, command: Command):
        command.execute()
        self.history.append(command)

    def undo_last(self):
        if self.history:
            command = self.history.pop()
            command.undo()

# Client
light = Light()
remote = RemoteControl()

on_cmd = LightOnCommand(light)
off_cmd = LightOffCommand(light)

remote.execute_command(on_cmd)
remote.execute_command(off_cmd)
remote.undo_last()  # Undoes the LightOffCommand
remote.undo_last()  # Undoes the LightOnCommand

File: 05-behavioral-patterns/03-command/03-command-undo.go
package main

import "fmt"

// Command interface
type Command interface {
	Execute()
	Undo()
}

// Receiver
type Light struct{}

func (l *Light) On() {
	fmt.Println("Light is ON")
}

func (l *Light) Off() {
	fmt.Println("Light is OFF")
}

// Concrete Commands
type LightOnCommand struct {
	light *Light
}

func (c *LightOnCommand) Execute() {
	c.light.On()
}

func (c *LightOnCommand) Undo() {
	c.light.Off()
}

type LightOffCommand struct {
	light *Light
}

func (c *LightOffCommand) Execute() {
	c.light.Off()
}

func (c *LightOffCommand) Undo() {
	c.light.On()
}

// Invoker
type RemoteControl struct {
	history []Command
}

func (r *RemoteControl) ExecuteCommand(c Command) {
	c.Execute()
	r.history = append(r.history, c)
}

func (r *RemoteControl) UndoLast() {
	if len(r.history) == 0 {
		return
	}
	last := r.history[len(r.history)-1]
	r.history = r.history[:len(r.history)-1]
	last.Undo()
}

// Client
func main() {
	light := &Light{}
	remote := &RemoteControl{}

	onCmd := &LightOnCommand{light}
	offCmd := &LightOffCommand{light}

	remote.ExecuteCommand(onCmd)
	remote.ExecuteCommand(offCmd)
	remote.UndoLast() // Undo LightOff
	remote.UndoLast() // Undo LightOn
}
